aptos-move/e2e-testsuite/src/tests/invariant_violation.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_language_e2e_tests::{common_transactions::peer_to_peer_txn, executor::FakeExecutor};
+use aptos_types::{
+    account_address::AccountAddress,
+    on_chain_config::FeatureFlag,
+    transaction::{ExecutionStatus, TransactionStatus},
+    vm_status::DiscardedVMStatus,
+};
+use move_core_types::{value::MoveValue, vm_status::StatusCode};
+
+#[test]
+fn invariant_violation_error() {
+    let _scenario = fail::FailScenario::setup();
+    fail::cfg(
+        "move_adapter::execute_script_or_entry_function",
+        "100%return",
+    )
+    .unwrap();
+
+    ::aptos_logger::Logger::init_for_testing();
+
+    let mut executor = FakeExecutor::from_head_genesis();
+
+    // create and publish a sender with 1_000_000 coins and a receiver with 100_000 coins
+    let sender = executor.create_raw_account_data(1_000_000, 10);
+    let receiver = executor.create_raw_account_data(100_000, 10);
+    executor.add_account_data(&sender);
+    executor.add_account_data(&receiver);
+
+    let transfer_amount = 1_000;
+    let txn = peer_to_peer_txn(sender.account(), receiver.account(), 10, transfer_amount, 0);
+
+    // execute transaction
+    let output = executor.execute_transaction(txn.clone());
+
+    // CHARGE_INVARIANT_VIOLATION enabled at genesis so this txn is kept.
+    assert_eq!(
+        output.status(),
+        &TransactionStatus::Keep(ExecutionStatus::MiscellaneousError(Some(
+            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR
+        ))),
+    );
+
+    // Disable the CHARGE_INVARIANT_VIOLATION flag.
+    executor.exec("features", "change_feature_flags", vec![], vec![
+        MoveValue::Signer(AccountAddress::ONE)
+            .simple_serialize()
+            .unwrap(),
+        MoveValue::Vector(vec![]).simple_serialize().unwrap(),
+        MoveValue::Vector(vec![MoveValue::U64(
+            FeatureFlag::CHARGE_INVARIANT_VIOLATION as u64,
+        )])
+        .simple_serialize()
+        .unwrap(),
+    ]);
+
+    let output = executor.execute_transaction(txn);
+
+    // With CHARGE_INVARIANT_VIOLATION disabled this transaction will be discarded.
+    assert_eq!(
```

Comment:
Here's the test @movekevin

Reply:
This only tests the output though and not the gas charge

Reply:
If it is discarded then it will definitelty not charge for gas right?

Reply:
Yeah but just adding an explicit check that gas has been deducted from the account (balance is less than in the beginning) provides absolute absolute confidence